<style type="text/css" media="screen">
/*
.nodes-image {
	margin:-100;
}
*/	
@import url("//maxcdn.bootstrapcdn.com/font-awesome/4.4.0/css/font-awesome.min.css");

.imageblock .content img, .image img {max-width: 900px;max-height: 300px;}
.deck h3, .deck h4 {display: block !important;margin-bottom:8px;margin-top:5px;}
.listingblock {margin:8px;}
.pull-bottom {position:relative;bottom:1em;}
.admonitionblock td.icon [class^="fa icon-"]{font-size:2.5em;text-shadow:1px 1px 2px rgba(0,0,0,.5);cursor:default}
.admonitionblock td.icon .icon-note:before{content:"\f05a";color:#19407c}
.admonitionblock td.icon .icon-tip:before{content:"\f0eb";text-shadow:1px 1px 2px rgba(155,155,0,.8);color:#111}
.admonitionblock td.icon .icon-warning:before{content:"\f071";color:#bf6900}
.admonitionblock td.icon .icon-caution:before{content:"\f06d";color:#bf3400}
.admonitionblock td.icon .icon-important:before{content:"\f06a";color:#bf0000}
.admonitionblock.note.speaker { display:none; }
</style>
<style type="text/css" media="screen">
/* #editor.maximize-editor .CodeMirror-code { font-size:24px; line-height:26px; } */
</style>
<article class="guide" ng-controller="AdLibDataController">
  <carousel class="deck container-fluid">
    <!--slide class="row-fluid">
      <div class="col-sm-3">
        <h3>HOW TO QUERY</h3>
        <p class="lead">Information</p>
			<!dl>
				
				
				<dt>author"</dt><dd>NÃºria Queralt Rosinach</dd>
				
				
			</dl>
		</div>
      <div class="col-sm-9">
        <figure>
          <img style="width:300px" src=""/>
        </figure>
      </div>
    </slide-->
    


<slide class="row-fluid">
  <div class="col-sm-12">
    <h3>HOW TO QUERY</h3>
    <br/>
    <div>
      


   <h4>Exploring with Neo4j</h4>
   <div class="paragraph">
<p><a href="https://neo4j.com/">Neo4j</a> is a graph database management system developed by Neo Technology, Inc. A <a href="https://en.wikipedia.org/wiki/Graph_database">graph database</a> is a database that uses graph structures for semantic queries with nodes, edges and properties to represent and store data. <a href="https://en.wikipedia.org/wiki/Cypher_Query_Language">Cypher</a> is the query language for Neo4j&#8217;s graphs. To follow this guide you should be able to read and understand Cypher queries.</p>
</div>
<div class="paragraph">
<p>You are currently interacting with the NGLY1 knowledge graph through the Neo4j browser, which provides a web interface to the database where you can run queries and explore the connected knowledge. To explore the whole NGLY1 knowledge graph, click on the following block to copy-paste the query into the shell and run it:</p>
</div>
<div class="listingblock">
<div class="content">
<pre mode="cypher"  class="highlight pre-scrollable programlisting cm-s-neo code runnable standalone-example ng-binding" data-lang="cypher" lang="cypher"><!--code class="cypher language-cypher"-->MATCH (n) RETURN * LIMIT 100<!--/code--></pre>
</div>
</div>
<div class="paragraph">
<p><br></p>
</div>
<div class="admonitionblock important">
<table>
<tr>
<td class="icon">
<i class="fa icon-important" title="Important"></i>
</td>
<td class="content">
Click the top right <strong>black&amp;white</strong> arrow to go to the next page
</td>
</tr>
</table>
</div>
	</div>
  </div>
</slide>



<slide class="row-fluid">
  <div class="col-sm-12">
    <h3>THE SEMANTICS OF THE NGLY1 DEFICIENCY KNOWLEDGE GRAPH v3.2</h3>
    <br/>
    <div>
      <div class="paragraph">
<p>In this page we provide the data model of our semantic graph and some example queries to retrieve data model information from the graph.</p>
</div>



   <h4>Data model</h4>
   <div class="paragraph">
<p>The data model is the abstraction of how knowledge is represented in the graph and it is necessary to run more interesting queries. The following picture shows the data model or schema for NGLY1 Knowledge Graph v3.0. The semantics of the concepts in the graph rely on the semantic types that build the <a href="https://www.nlm.nih.gov/pubs/factsheets/umlssemn.html">UMLS Semantic Network</a>. To check the definition of each semantic type abbreviation see the <a href="https://metamap.nlm.nih.gov/Docs/SemGroups_2013.txt">UMLS Semantic Groups definitions</a>.</p>
</div>
<div id="img-schema" class="imageblock" style="float: center;">
<div class="content">
<a class="image" href="https://www.dropbox.com/s/6m1rfik5qqftyvi/ngly1kg_metagraph_2019-03-12.png?dl=0"><img src="https://www.dropbox.com/s/6m1rfik5qqftyvi/ngly1kg_metagraph_2019-03-12.png?dl=1" alt="ngly1kg metagraph 2019 03 12"></a>
</div>
<div class="title">Figure 1. NGLY1 Deficiency Knowledge Graph v3.2 data model</div>
</div>
<div class="paragraph">
<p><br></p>
</div>



   <h4>Retrieve information about the data model</h4>
   <div class="paragraph">
<p>To retrieve the explicit semantics to use for node types and its attributes in the graph, run the following query:</p>
</div>
<div class="listingblock">
<div class="content">
<pre mode="cypher"  class="highlight pre-scrollable programlisting cm-s-neo code runnable standalone-example ng-binding" data-lang="cypher" lang="cypher"><!--code class="cypher language-cypher"-->MATCH (n)
RETURN DISTINCT labels(n) AS node_type,
count(*) AS NumberOfEntities, reduce(keys = [], keys_n in collect(keys(n)) | keys + filter(k in keys_n WHERE NOT k IN keys)) AS EntityAttributes
ORDER BY NumberOfEntities DESC<!--/code--></pre>
</div>
</div>
<div class="paragraph">
<p>To retrieve the explicit semantics to use for edge types and their attributes, run the following query:</p>
</div>
<div class="listingblock">
<div class="content">
<pre mode="cypher"  class="highlight pre-scrollable programlisting cm-s-neo code runnable standalone-example ng-binding" data-lang="cypher" lang="cypher"><!--code class="cypher language-cypher"-->MATCH ()-[r]-()
RETURN DISTINCT type(r) AS relationship_type, r.property_label AS relationship_label,
count(*) AS NumberOfRelationships, reduce(keys = [], keys_r in collect(keys(r)) | keys + filter(k in keys_r WHERE NOT k IN keys)) AS EntityAttributes
ORDER BY NumberOfRelationships DESC<!--/code--></pre>
</div>
</div>
<div class="paragraph">
<p>To retrieve what is related and how, run the following query:</p>
</div>
<div class="listingblock">
<div class="content">
<pre mode="cypher"  class="highlight pre-scrollable programlisting cm-s-neo code runnable standalone-example ng-binding" data-lang="cypher" lang="cypher"><!--code class="cypher language-cypher"-->MATCH (n)
MATCH (n)-[r]-&gt;(m)
WITH n, type(r) AS via, m
RETURN labels(n) AS from, via, labels(m) AS to, count(*) AS freq
ORDER BY freq DESC<!--/code--></pre>
</div>
</div>
<div class="paragraph">
<p><br></p>
</div>
<div class="admonitionblock important">
<table>
<tr>
<td class="icon">
<i class="fa icon-important" title="Important"></i>
</td>
<td class="content">
Click the top right <strong>black&amp;white</strong> arrow to go to the next page
</td>
</tr>
</table>
</div>
	</div>
  </div>
</slide>



<slide class="row-fluid">
  <div class="col-sm-12">
    <h3>HINTS</h3>
    <br/>
    <div>
      


   <h4>How to extract hypotheses</h4>
   <div class="paragraph">
<p>Hypotheses are derived from exploring the evidence connecting a source-target pair of nodes. To explore this evidence we traverse the NGLY1 Knowledge Graph running queries. Hypotheses are extracted querying for semantic paths connecting source and target nodes. In the next pages, we show examples of hypotheses extracted by specific queries between two nodes. Each page is dedicated to a source-target pair and we provide several query templates to illustrate and guide the user how to query the graph to extract hypotheses. To access these pairwise hypotheses pages the user has to click the b/n arrow on the top-right part of the web. For instance, if we click this arrow we access the first page that is dedicated to the NGLY1 Deficiency-NGLY1 pair. In this page we provide examples on how to run open queries to retrieve hypotheses of path length not longer than 2 edges, i.e. without specifying a explicit pattern connecting two nodes, using or not filters.</p>
</div>



   <h4>How to query the graph</h4>
   <div class="paragraph">
<p>To query the graph we have to use the Cypher query language. This language allows us to query nodes, node attributes, relationships between nodes and attributes of these relationships. We can perform open queries (without a specific query template), which are usually computationally less efficient for large graphs. Alternatively, we can be more specific and search for explicit nodes and connection patterns using semantics and a query template, which makes queries more efficient in large data graphs. The data model shows how things are meaningfully represented and related to each other and it is the base to use semantics in the query. To illustrate how to query for specific semantic connections the NGLY1 Knowledge Graph, we provide some examples in Cypher using specific query templates. We do not provide a tutorial for the Cypher language. To learn how to query using Cypher the user should refer to external sources like <a href="https://neo4j.com/developer/cypher-query-language/">The Neo4j&#8217;s introduction to Cypher</a></p>
</div>
<div class="paragraph">
<p><br></p>
</div>
<div class="admonitionblock important">
<table>
<tr>
<td class="icon">
<i class="fa icon-important" title="Important"></i>
</td>
<td class="content">
Click the top right <strong>black&amp;white</strong> arrow to go to the next page
</td>
</tr>
</table>
</div>
	</div>
  </div>
</slide>



<slide class="row-fluid">
  <div class="col-sm-12">
    <h3>QUERIES FOR GRAPH TRAVERSAL ANALYSIS. QUERY EXAMPLE: NGLY1 --- AQP1 PATHS</h3>
    <br/>
    <div>
      <div class="paragraph">
<p>To solve a question we explore and search the review by querying the graph. Let&#8217;s develop the whole process with a query example.</p>
</div>



   <h4>Open query: extraction of metapaths</h4>
   <div class="paragraph">
<p><em>Open query</em></p>
</div>
<div class="listingblock">
<div class="content">
<pre mode="cypher"  class="highlight pre-scrollable programlisting cm-s-neo code runnable standalone-example ng-binding" data-lang="cypher" lang="cypher"><!--code class="cypher language-cypher"-->MATCH path=(source:GENE)-[i1:`RO:0002434`]-&gt;(:GENE)-[*..3]-(target:GENE)

WHERE source.id = 'FlyBase:FBgn0033050' AND target.id = 'HGNC:633' AND ALL(x IN nodes(path) WHERE single(y IN nodes(path) WHERE y = x))

WITH path,i1,

[n IN nodes(path) WHERE n.preflabel IN ['cytoplasm','cytosol','nucleus','metabolism','membrane','protein binding','visible','viable','phenotype']] AS nodes_marked

WHERE size(nodes_marked) = 0 AND toLower(i1.reference_uri) contains 'pubmed/29346549'

RETURN count(distinct path) as paths<!--/code--></pre>
</div>
</div>
<div class="paragraph">
<p><em>Metapaths</em></p>
</div>
<div class="listingblock">
<div class="content">
<pre mode="cypher"  class="highlight pre-scrollable programlisting cm-s-neo code runnable standalone-example ng-binding" data-lang="cypher" lang="cypher"><!--code class="cypher language-cypher"-->MATCH path=(source:GENE)-[i1:`RO:0002434`]-&gt;(:GENE)-[*..3]-(target:GENE)

WHERE source.id = 'FlyBase:FBgn0033050' AND target.id = 'HGNC:633' AND ALL(x IN nodes(path) WHERE single(y IN nodes(path) WHERE y = x))

WITH path,i1,

[n IN nodes(path) WHERE n.preflabel IN ['cytoplasm','cytosol','nucleus','metabolism','membrane','protein binding','visible','viable','phenotype']] AS nodes_marked

WHERE size(nodes_marked) = 0 AND toLower(i1.reference_uri) contains 'pubmed/29346549'

RETURN DISTINCT extract (x in rels(path) | type(x)) as types, extract (n in nodes(path) | labels(n)) as labels, length(path) as mp_length, count(distinct path) as paths

ORDER BY mp_length, paths DESC<!--/code--></pre>
</div>
</div>



   <h4>Query template</h4>
   <div class="paragraph">
<p>To structure the query in Cypher, it is important to first formulate the question in natural language and then translate it in a query template.</p>
</div>



   <h4>Query in Cypher</h4>
   <div class="paragraph">
<p>Once we have a clear idea of the query template, it is easier to formulated in a structured way in Cypher.</p>
</div>
<div class="paragraph">
<p><em>Query template</em></p>
</div>
<div class="listingblock">
<div class="content">
<pre mode="cypher"  class="highlight pre-scrollable programlisting cm-s-neo code runnable standalone-example ng-binding" data-lang="cypher" lang="cypher"><!--code class="cypher language-cypher"-->MATCH path=(source:GENE)-[i1:`RO:0002434`]-&gt;(:GENE)-[:`RO:0002206`]-(:ANAT)-[:`RO:0002206`]-(:GENE)-[:`RO:0002434`]-(target:GENE)

WHERE source.id = 'FlyBase:FBgn0033050' AND target.id = 'HGNC:633' AND ALL(x IN nodes(path) WHERE single(y IN nodes(path) WHERE y = x))

WITH path,i1,

[n IN nodes(path) WHERE n.preflabel IN ['cytoplasm','cytosol','nucleus','metabolism','membrane','protein binding','visible','viable','phenotype']] AS nodes_marked

WHERE size(nodes_marked) = 0 AND toLower(i1.reference_uri) contains 'pubmed/29346549'

RETURN count(distinct path) as paths<!--/code--></pre>
</div>
</div>



   <h4>Results</h4>
   <div class="paragraph">
<p><em>Table</em></p>
</div>
<div class="listingblock">
<div class="content">
<pre mode="cypher"  class="highlight pre-scrollable programlisting cm-s-neo code runnable standalone-example ng-binding" data-lang="cypher" lang="cypher"><!--code class="cypher language-cypher"-->MATCH path=(source:GENE)-[i1:`RO:0002434`]-&gt;(:GENE)-[:`RO:0002206`]-(:ANAT)-[:`RO:0002206`]-(:GENE)-[:`RO:0002434`]-(target:GENE)

WHERE source.id = 'FlyBase:FBgn0033050' AND target.id = 'HGNC:633' AND ALL(x IN nodes(path) WHERE single(y IN nodes(path) WHERE y = x))

WITH path,i1,

[n IN nodes(path) WHERE n.preflabel IN ['cytoplasm','cytosol','nucleus','metabolism','membrane','protein binding','visible','viable','phenotype']] AS nodes_marked

WHERE size(nodes_marked) = 0 AND toLower(i1.reference_uri) contains 'pubmed/29346549'

RETURN DISTINCT extract(x in rels(path) | x.property_label) AS types, extract(n in nodes(path) | n.name) AS labels, length(path) AS mp_length, count(distinct path) AS paths

ORDER BY mp_length, paths DESC<!--/code--></pre>
</div>
</div>



   <h4>Refine query template</h4>
   


   <h4>Query in Cypher</h4>
   <div class="paragraph">
<p>The query in Cypher is:</p>
</div>
<div class="paragraph">
<p><em>Refine query template</em></p>
</div>
<div class="listingblock">
<div class="content">
<pre mode="cypher"  class="highlight pre-scrollable programlisting cm-s-neo code runnable standalone-example ng-binding" data-lang="cypher" lang="cypher"><!--code class="cypher language-cypher"-->MATCH path=(source:GENE)-[i1:`RO:0002434`]-&gt;(:GENE)-[:`RO:HOM0000017`]-(:GENE)-[:`RO:0002434`]-(:GENE)-[:`RO:0002434`]-(target:GENE)

WHERE source.id = 'FlyBase:FBgn0033050' AND target.id = 'HGNC:633' AND ALL(x IN nodes(path) WHERE single(y IN nodes(path) WHERE y = x))

WITH path,i1,

[n IN nodes(path) WHERE n.preflabel IN ['cytoplasm','cytosol','nucleus','metabolism','membrane','protein binding','visible','viable','phenotype']] AS nodes_marked

WHERE size(nodes_marked) = 0 AND toLower(i1.reference_uri) contains 'pubmed/29346549'

RETURN count(distinct path) as paths<!--/code--></pre>
</div>
</div>



   <h4>Results</h4>
   <div class="paragraph">
<p>The results appear below the command shell as a graph or as a table. Results are only exportable as files: graphic (PNG, SVG, ..) or structured (CVS, ..) formats.</p>
</div>
<div class="admonitionblock warning">
<table>
<tr>
<td class="icon">
<i class="fa icon-warning" title="Warning"></i>
</td>
<td class="content">
The current Neo4j Browser does not provide <code>session</code> services, thus there is not option to save any session.
</td>
</tr>
</table>
</div>
<div class="paragraph">
<p><br></p>
</div>
<div class="admonitionblock important">
<table>
<tr>
<td class="icon">
<i class="fa icon-important" title="Important"></i>
</td>
<td class="content">
Click the top right <strong>black&amp;white</strong> arrow to go to the next page
</td>
</tr>
</table>
</div>
	</div>
  </div>
</slide>



<slide class="row-fluid">
  <div class="col-sm-12">
    <h3>QUERIES FOR TRANSCRIPTOME ANALYSIS</h3>
    <br/>
    <div>
      


   <h4>Query for common TFs</h4>
   <div class="listingblock">
<div class="content">
<pre mode="cypher"  class="highlight pre-scrollable programlisting cm-s-neo code runnable standalone-example ng-binding" data-lang="cypher" lang="cypher"><!--code class="cypher language-cypher"-->MATCH path=(source:GENE)-[i1:`RO:0002434`]-&gt;(ortho:GENE)-[iortho]-(gene:GENE)&lt;-[i2:`RO:0002434`]-(tf:GENE)

WHERE source.id = 'FlyBase:FBgn0033050' AND ALL(x IN nodes(path) WHERE single(y IN nodes(path) WHERE y = x))

WITH path,i1,iortho,i2,tf,gene,

[n IN nodes(path) WHERE n.preflabel IN ['cytoplasm','cytosol','nucleus','metabolism','membrane','protein binding','visible','viable','phenotype']] AS nodes_marked

WHERE size(nodes_marked) = 0

AND toLower(i1.reference_uri) =~ '.*pubmed/29346549.*'

AND toLower(iortho.property_label) CONTAINS 'orthology'

AND toLower(i2.reference_supporting_text) =~ '.*tftargets.*|.*msigdb.*'

RETURN DISTINCT tf.id as TF_id, tf.preflabel as TF_symbol, tf.name as TF_name, tf.description as TF_description, count(distinct gene.preflabel) as Freq

ORDER BY Freq DESC<!--/code--></pre>
</div>
</div>
<div class="paragraph">
<p><em>Summarize frequency and target gene sets for each TF</em></p>
</div>
<div class="listingblock">
<div class="content">
<pre mode="cypher"  class="highlight pre-scrollable programlisting cm-s-neo code runnable standalone-example ng-binding" data-lang="cypher" lang="cypher"><!--code class="cypher language-cypher"-->MATCH path=(source:GENE)-[i1:`RO:0002434`]-&gt;(o:GENE)-[iortho]-(g:GENE)&lt;-[i2:`RO:0002434`]-(tf:GENE)

WHERE source.id = 'FlyBase:FBgn0033050' AND ALL(x IN nodes(path) WHERE single(y IN nodes(path) WHERE y = x))

WITH path,i1,iortho,i2,g,tf,

[n IN nodes(path) WHERE n.preflabel IN ['cytoplasm','cytosol','nucleus','metabolism','membrane','protein binding','visible','viable','phenotype']] AS nodes_marked

WHERE size(nodes_marked) = 0

AND toLower(i1.reference_uri) =~ '.*pubmed/29346549.*'

AND toLower(iortho.property_label) CONTAINS 'orthology'

AND toLower(i2.reference_supporting_text) =~ '.*tftargets.*|.*msigdb.*'

RETURN DISTINCT tf.preflabel as TF_symbol, tf.name as TF_name, COLLECT(DISTINCT g.preflabel) as geneset, count(distinct g.preflabel) as genes

ORDER BY genes DESC<!--/code--></pre>
</div>
</div>
<div class="paragraph">
<p><br></p>
</div>



   <h4>Query for TF Pathways</h4>
   <div class="imageblock" style="text-align: center;">
<div class="content">
<img src="https://www.dropbox.com/s/kvty5e0fuoqdadh/query3_topology_v3.2.png?dl=1" alt="query3 topology v3.2" width="5000" height="1000">
</div>
</div>
<div class="listingblock">
<div class="content">
<pre mode="cypher"  class="highlight pre-scrollable programlisting cm-s-neo code runnable standalone-example ng-binding" data-lang="cypher" lang="cypher"><!--code class="cypher language-cypher"-->MATCH path=(source:GENE)-[i1:`RO:0002434`]-&gt;(o:GENE)-[iortho]-(g:GENE)&lt;-[i2:`RO:0002434`]-(tf:GENE), (g:GENE)-[i]-(pw:PHYS)

WHERE source.id = 'FlyBase:FBgn0033050' AND ALL(x IN nodes(path) WHERE single(y IN nodes(path) WHERE y = x))

WITH path,i1,iortho,i2,g,tf,i,pw,

[n IN nodes(path) WHERE n.preflabel IN ['cytoplasm','cytosol','nucleus','metabolism','membrane','protein binding','visible','viable','phenotype']] AS nodes_marked

WHERE size(nodes_marked) = 0

AND toLower(i1.reference_uri) =~ '.*pubmed/29346549.*'

AND toLower(iortho.property_label) CONTAINS 'orthology'

AND toLower(i2.reference_supporting_text) =~ '.*tftargets.*|.*msigdb.*'

AND toLower(i.property_label) &lt;&gt; 'enables'

RETURN DISTINCT tf.preflabel as TF_symbol, tf.name as TF_name,
                collect(DISTINCT g.preflabel) as geneset, count(distinct g.preflabel) as genes,
                collect(DISTINCT pw.preflabel) as pathway, count(distinct pw.preflabel) as pathways
                                                                                                                                            ORDER BY genes DESC<!--/code--></pre>
</div>
</div>
<div class="paragraph">
<p><br></p>
</div>



   <h4>Query for TF GO annotations: Cellular function</h4>
   <div class="listingblock">
<div class="content">
<pre mode="cypher"  class="highlight pre-scrollable programlisting cm-s-neo code runnable standalone-example ng-binding" data-lang="cypher" lang="cypher"><!--code class="cypher language-cypher"-->MATCH path=(source:GENE)-[i1:`RO:0002434`]-&gt;(o:GENE)-[iortho]-(g:GENE)&lt;-[i2:`RO:0002434`]-(tf:GENE), (g:GENE)-[i]-(pw:PHYS)

WHERE source.id = 'FlyBase:FBgn0033050' AND ALL(x IN nodes(path) WHERE single(y IN nodes(path) WHERE y = x))

WITH path,i1,iortho,i2,g,tf,i,pw,

[n IN nodes(path) WHERE n.preflabel IN ['cytoplasm','cytosol','nucleus','metabolism','membrane','protein binding','visible','viable','phenotype']] AS nodes_marked

WHERE size(nodes_marked) = 0

AND toLower(i1.reference_uri) =~ '.*pubmed/29346549.*'

AND toLower(iortho.property_label) CONTAINS 'orthology'

AND toLower(i2.reference_supporting_text) =~ '.*tftargets.*|.*msigdb.*'

AND toLower(i.property_label) = 'enables'

RETURN DISTINCT tf.preflabel as TF_symbol, tf.name as TF_name,
                collect(DISTINCT g.preflabel) as geneset, count(distinct g.preflabel) as genes,
                collect(DISTINCT pw.preflabel) as cellular_function, count(distinct pw.preflabel) as cellular_functions
                                                                                                                                        ORDER BY genes DESC<!--/code--></pre>
</div>
</div>
	</div>
  </div>
</slide>
  </carousel>
</article>