:author: NÃºria Queralt Rosinach
:twitter: nqueralt
:tags: NGLY1 Deficiency, NGLY1, Hypothesis Generation
:neo4j-version: 3.0.3


== INTRODUCTION


=== NGLY1 DEFICIENCY KNOWLEDGE GRAPH


**Detail the knowledge content in the structured reviewed**

The NGLY1 Deficiency Knowledge Graph is a network of biology, medicine, and chemistry. It contains the current knowledge about the NGLY1 deficiency-CDDG disease curated from the scientific literature and from the biomedical knowledge. The current version 3.2 contains 9,365 nodes of 8 types and 237,027 relationships of 32 types.

We created the NGLY1 Deficiency Knowledge Graph to support translational research on the NGLY1 deficiency. We reviewed and structured information around the clinical profile, the pathophysiology and the pharmacology of the NGLY1 deficiency to help researchers to discover evidence-based explanations for their hypotheses. The third release of the graph organizes curated information from human and animal models 1) manually extracted from the literature; 2) retrieved from https://monarchinitiative.org/[the Monarch Initiative]; 3) transcription factors and their targeted genes from several sources in tftargets and MSigDB; 4) manually contributed by the community from http://100.25.145.12:8181/wiki/Main_Page[the NGLY1 Deficiency wikibase platform]; and integrates it with an experimental expression data from mouse. 


=== WHAT KIND OF QUESTIONS WE CAN DO

**Highlight what we can learn from the knowledge reviewed**

The user can query the knowledge graph asking for paths between entities to discover new hypotheses or extracting data that can shed light or give new ideas. +



=== STRUCTURE OF THE GUIDE


This guide is structured with query pages in the next section, where the guide shows an explicit question and develops a type of query we can formulate to inspect hypotheses from the knowledge in the review. In the last section, the guide develops how to query the graph with some hints and examples to aid the new user. To see how to retrieve connections through regulation between NGLY1-AQP1 go to:

* _Slide 2_: *DATA MODEL*

* _Slide 3_: *QUERY 1: HYPOTHESIS DISCOVERY RELATING NGLY1 - AQP1* 

* _Slide 4_: *QUERY 2: HYPOTHESIS DISCOVERY RELATING AQP1 - DISEASE PHENOTYPES* 

* _Slide 5_: *LAST REMARKS*


IMPORTANT:  Click the top right **black&white** arrow to go to the next page


== DATA MODEL

=== DESCRIPTION
The picture shows the data model or schema or meta-graph for NGLY1 Knowledge Graph v3.2. The semantics of the concepts in the graph rely on the semantic types that build the https://www.nlm.nih.gov/pubs/factsheets/umlssemn.html[UMLS Semantic Network]. To check the definition of each semantic type abbreviation see the https://metamap.nlm.nih.gov/Docs/SemGroups_2013.txt[UMLS Semantic Groups definitions].

image::https://www.dropbox.com/s/6p1zykpuexslx8e/ngly1_graph_schema.png?dl=1[height=300,float=center]

For an interactive description of the data model, run the following query:

[source,cypher]
----
CALL db.schema()
----


=== ORIGINAL DATA SOURCES

Description of the orginal data



== QUERY 1: HYPOTHESIS DISCOVERY RELATING NGLY1 - AQP1 


Researchers have observed an association between NGLY1 human gene and AQP1 human gene at expression level. Their hypothesis is that both genes are linked through transcriptional regulation and they want to search the knowledge and the expression data integrated in the review to find potential mechanisms. In particular, they want to query the review to extract regulatory paths through genetic interactions that link NGLY1 and AQP1. +




=== STYLE THE GRAPH


We first style the graph, before playing with it. To load the NGLY1 Deficiency graph style into the Neo4j Browser, run the following command by clicking the box, which will populate the shell, and hitting ```enter``` (or ```return``` in some keyboards). This command only needs to be run once per web browser.


[source,cypher]
----
:style http://100.25.145.12:8001/guides/style.grass
----


TIP:  The user can set up the style of the graph by itself locally in its browser. If you click on any label or relationship above the graph visualization, you can then choose its styling in the area below the graph. Colors, sizes, and captions are selectable from there. Click http://guides.neo4j.com/browser.html[Styling Neo4j Browser Visualization] for more details. 


---

=== QUERY

In particular researchers are interested in hypothesis that link both entities through gene interactions. To mine the gap of knowledge we will explore for hypothesis that link the perturbed transcriptomic expression on fly model of NGLY1 Deficiency through regulatory TF-target gene interactions to human AQP1. Since we are translating fly biology to human biology we will explore paths with a 2-hop `gene-interacts with` flexibility. 

==== Define the query template


The first think to do is to formulate the query template to answer this question:

image::https://www.dropbox.com/s/qypkrn1l50p83n6/query1_topology_v3.2.png?dl=1[height=100, width=5000, align="center"]

In this query, we specify relations between genes to be of two types: 'interacts with' that will be explicitly either of regulatory origin once filtered by source of provenance to the databases *tftargets* and *MSigDB* or of gene expression origin once filtered by reference to the *PubMed Identifier*, where the experimental fly transcriptome data set was published, or a gene-gene interaction if none of these restrictions applied; and 'in 1 to 1 orthology relationship with' that relates to direct orthology between genes.


==== Formulate the query in Cypher


Once we have a clear query template we can create the query in Cypher to search for mechanistic links based on gene interactions and regulation edges. The following query traverses the graph to find these kind of links in the review:

[%autofit]
----
MATCH path=(source:GENE)-[i1:`RO:0002434`]->(rna:GENE)-[i2:`RO:HOM0000020`]-(ortholog_rna:GENE)-[i3:`RO:0002434`]-(tf:GENE)-[i4:`RO:0002434`]->(target:GENE)

WHERE source.id = 'FlyBase:FBgn0033050' AND target.id = 'HGNC:633' AND ALL(x IN nodes(path) WHERE single(y IN nodes(path) WHERE y = x))

WITH path,

[n IN nodes(path) WHERE n.preflabel IN ['cytoplasm','cytosol','nucleus','metabolism','membrane','protein binding','visible','viable','phenotype']] AS nodes_marked,

[r IN relationships(path) WHERE toLower(r.reference_supporting_text) =~ '.*tftargets.*|.*msigdb.*'] AS interactions

WHERE size(nodes_marked) = 0 AND size(interactions) <> 0

AND toLower(i1.reference_uri) CONTAINS 'pubmed/29346549'

AND toLower(i4.reference_supporting_text) =~ '.*tftargets.*|.*msigdb.*'

RETURN path
----


Let's explain this query statement by statement. First, we look for mechanisms that match the query template above, where we define gene interactions with the node type _GENE_ and the edge types  _RO:0002434_ (interacts with) and _RO:HOM0000020_ (in 1 to 1 orthology relationship with):


 MATCH path=(source:GENE)-[i1:`RO:0002434`]->(rna:GENE)-[i2:`RO:HOM0000020`]-(ortholog_rna:GENE)-[i3:`RO:0002434`]-(tf:GENE)-[i4:`RO:0002434`]->(target:GENE)


We define the source and target nodes with the identifier (id). The source is Pngl (FlyBase:FBgn0033050), which is the fly ortholog of the human gene NGLY1, (HGNC:17646) and the target is the human gene AQP1 (HGNC:633)


 WHERE source.id = 'FlyBase:FBgn0033050' AND target.id = 'HGNC:633'


and we impose that every every node instance is traversed only once in the path.


 AND ALL(x IN nodes(path) WHERE single(y IN nodes(path) WHERE y = x))


We extract some features from each path to then filter them


 WITH path,


We mark nodes with too general meaning


 [ n IN nodes(path) WHERE n.preflabel IN ['cytoplasm','cytosol','nucleus','metabolism','membrane','protein binding','visible','viable','phenotype'] ] AS nodes_marked,


We mark regulatory edges through the "reference_supporting_text" edge provenance attribute


 [ r IN relationships(path) WHERE toLower(r.reference_supporting_text) =~ '.*tftargets.*|.*msigdb.*' ] AS interactions


We filter paths that only content specific node instances and at least have one regulatory edge


 WHERE size(nodes_marked) = 0 AND size(interactions) <> 0


We filter paths that the first edge, named in the path pattern as 'i1', is a transcriptomic interaction from the experimental dataset published in the reference id PMID:29346549. We filter it by publication provenance using the `reference_uri` edge attribute


 AND toLower(i1.reference_uri) CONTAINS 'pubmed/29346549'


We filter paths that the fourth edge, named as 'i4', is a TF --> GENE regulatory interaction. We define a specific directionality of the interaction to keep the TF function towards AQP1. We filter it by source provenance using the `reference_supporting_text` edge attribute


 AND toLower(i4.reference_supporting_text) =~ '.*tftargets.*|.*msigdb.*'


We retrieve paths


 RETURN path


{sp}+

---


=== RESULTS


Results can be vizualized in a summary table of a list of properties calculated from all paths or in a graph exposing all the paths. In this case we can explore all paths directly in graph format that looks like:


image::https://www.dropbox.com/s/bbdg1orj7ztrvup/query1_paths_v3.2.png?dl=1[height=3000, width=10500]


If we want to retrieve a summary table with all gene interactors that are expressed gene interactors of Pngl and AQP1 Transcription Factors (TFs) ordered by the total number of paths they appear, the query should be:


 MATCH path=(source:GENE)-[i1:`RO:0002434`]->(rna:GENE)-[i2:`RO:HOM0000020`]-(ortholog_rna:GENE)-[i3:`RO:0002434`]-(tf:GENE)-[i4:`RO:0002434`]->(target:GENE)

 WHERE source.id = 'FlyBase:FBgn0033050' AND target.id = 'HGNC:633' AND ALL(x IN nodes(path) WHERE single(y IN nodes(path) WHERE y = x))

 WITH path,rna,tf,

 [n IN nodes(path) WHERE n.preflabel IN ['cytoplasm','cytosol','nucleus','metabolism','membrane','protein binding','visible','viable','phenotype']] AS nodes_marked,

 [r IN relationships(path) WHERE toLower(r.reference_supporting_text) =~ '.*tftargets.*|.*msigdb.*'] AS interactions

 WHERE size(nodes_marked) = 0 AND size(interactions) <> 0

 AND toLower(i1.reference_uri) CONTAINS 'pubmed/29346549'

 AND toLower(i4.reference_supporting_text) =~ '.*tftargets.*|.*msigdb.*'

 RETURN DISTINCT rna.name AS Expressed_gene_name, tf.name AS TF_name, count(distinct path) as Total_paths

 ORDER BY Total_paths DESC


Which results in the following table:


image::https://www.dropbox.com/s/v7i3ld0x3cup7nj/query1_table_v3.2.png?dl=1[height=1000, width=9000]


TIP: If the user wants to list ortholog rna genes, replace `rna` label by `ortholog_rna` in the query to refer to the third node position in the query template.

---


=== TRY IT!


Execute the query yourself and explore results.


==== Query 1: paths that relate NGLY1 and AQP1 through gene interactions


_Query network_


[source,cypher]
----
MATCH path=(source:GENE)-[i1:`RO:0002434`]->(rna:GENE)-[i2:`RO:HOM0000020`]-(ortholog_rna:GENE)-[i3:`RO:0002434`]-(tf:GENE)-[i4:`RO:0002434`]->(target:GENE)

WHERE source.id = 'FlyBase:FBgn0033050' AND target.id = 'HGNC:633' AND ALL(x IN nodes(path) WHERE single(y IN nodes(path) WHERE y = x))

WITH path,

[n IN nodes(path) WHERE n.preflabel IN ['cytoplasm','cytosol','nucleus','metabolism','membrane','protein binding','visible','viable','phenotype']] AS nodes_marked,

[r IN relationships(path) WHERE toLower(r.reference_supporting_text) =~ '.*tftargets.*|.*msigdb.*'] AS interactions

WHERE size(nodes_marked) = 0 AND size(interactions) <> 0

AND toLower(i1.reference_uri) CONTAINS 'pubmed/29346549'

AND toLower(i4.reference_supporting_text) =~ '.*tftargets.*|.*msigdb.*'

RETURN path
----


==== Query 2: summary table of gene interactors in the resulting paths


_Query network_


[source,cypher]
----
MATCH path=(source:GENE)-[i1:`RO:0002434`]->(rna:GENE)-[i2:`RO:HOM0000020`]-(ortholog_rna:GENE)-[i3:`RO:0002434`]-(tf:GENE)-[i4:`RO:0002434`]->(target:GENE)

WHERE source.id = 'FlyBase:FBgn0033050' AND target.id = 'HGNC:633' AND ALL(x IN nodes(path) WHERE single(y IN nodes(path) WHERE y = x))

WITH path,rna,tf,

[n IN nodes(path) WHERE n.preflabel IN ['cytoplasm','cytosol','nucleus','metabolism','membrane','protein binding','visible','viable','phenotype']] AS nodes_marked,

[r IN relationships(path) WHERE toLower(r.reference_supporting_text) =~ '.*tftargets.*|.*msigdb.*'] AS interactions

WHERE size(nodes_marked) = 0 AND size(interactions) <> 0

AND toLower(i1.reference_uri) CONTAINS 'pubmed/29346549'

AND toLower(i4.reference_supporting_text) =~ '.*tftargets.*|.*msigdb.*'

RETURN DISTINCT rna.name AS Expressed_gene_name, tf.name AS TF_name, count(distinct path) as Total_paths

ORDER BY Total_paths DESC
----



IMPORTANT:  Click the top right **black&white** arrow to go to the next page


== QUERY 2: HYPOTHESIS DISCOVERY RELATING AQP1 - DISEASE PHENOTYPES


Next question that the researcher may want to explore is the connection between the AQP1 and the NGLY1 Deficiency disease phenotypes.


---


=== QUERY


=== Define the query template


The first think to do is to formulate the query template to answer this question:


image::https://www.dropbox.com/s/ch261h9xwpu5zm5/query2_topology_v3.2.png?dl=1[height=600, width=2500, align="center"]

This query is in two steps.


=== Formulate the query in Cypher


---


=== RESULTS


The resulting paths can be visualized such as


image::https://www.dropbox.com/s/xh14124tnqz0noo/query2_paths_v3.2.png?dl=1[height=500, width=3000]


or can be summarized as a table such as


image::https://www.dropbox.com/s/vksf6gccraaf92g/query2_table_v3.2.png?dl=1[height=500, width=3000]


--- 


=== TRY IT!


==== Query 1


_Query network_


[source,cypher]
----
MATCH path=(source:GENE)-[i1:`RO:0002434`]->(rna:GENE)-[i2:`RO:HOM0000020`]-(ortholog_rna:GENE)-[i3:`RO:0002434`]-(tf:GENE)-[i4:`RO:0002434`]->(target:GENE)

WHERE source.id = 'FlyBase:FBgn0033050' AND target.id = 'HGNC:633' AND ALL(x IN nodes(path) WHERE single(y IN nodes(path) WHERE y = x))

WITH path,ortholog_rna,tf,

[n IN nodes(path) WHERE n.preflabel IN ['cytoplasm','cytosol','nucleus','metabolism','membrane','protein binding','visible','viable','phenotype']] AS nodes_marked,

[r IN relationships(path) WHERE toLower(r.reference_supporting_text) =~ '.*tftargets.*|.*msigdb.*'] AS interactions

WHERE size(nodes_marked) = 0 AND size(interactions) <> 0

AND toLower(i1.reference_uri) CONTAINS 'pubmed/29346549'

AND toLower(i4.reference_supporting_text) =~ '.*tftargets.*|.*msigdb.*'

WITH COLLECT(DISTINCT ortholog_rna.id) + COLLECT(DISTINCT tf.id) AS genes

MATCH path=(:DISO {id: 'DOID:0060728'})-->(ph:DISO)--(g:GENE)

WHERE g.id in genes AND ph.id CONTAINS 'HP:'

RETURN path
----


==== Query 2


_Query network_


[source,cypher]
----
MATCH path=(source:GENE)-[i1:`RO:0002434`]->(rna:GENE)-[i2:`RO:HOM0000020`]-(ortholog_rna:GENE)-[i3:`RO:0002434`]-(tf:GENE)-[i4:`RO:0002434`]->(target:GENE)

WHERE source.id = 'FlyBase:FBgn0033050' AND target.id = 'HGNC:633' AND ALL(x IN nodes(path) WHERE single(y IN nodes(path) WHERE y = x))

WITH path,ortholog_rna,tf,

[n IN nodes(path) WHERE n.preflabel IN ['cytoplasm','cytosol','nucleus','metabolism','membrane','protein binding','visible','viable','phenotype']] AS nodes_marked,

[r IN relationships(path) WHERE toLower(r.reference_supporting_text) =~ '.*tftargets.*|.*msigdb.*'] AS interactions

WHERE size(nodes_marked) = 0 AND size(interactions) <> 0

AND toLower(i1.reference_uri) contains 'pubmed/29346549'

AND toLower(i4.reference_supporting_text) =~ '.*tftargets.*|.*msigdb.*'

WITH COLLECT(DISTINCT ortholog_rna.id) + COLLECT(DISTINCT tf.id) AS genes

MATCH path=(:DISO {id: 'DOID:0060728'})-->(ph:DISO)--(g:GENE)

WHERE g.id in genes AND ph.id CONTAINS 'HP:'

RETURN DISTINCT ph.id AS phenotype_id, ph.preflabel AS phenotype_label, COLLECT(DISTINCT g.name) AS gene_id_list, count(DISTINCT g.id) AS genes

ORDER BY genes DESC
----


IMPORTANT:  Click the top right **black&white** arrow to go to the next page


== LAST REMARKS
**Last wrapped up or conclusion section**


=== HYPOTHESIS DISCOVERY ON THE NGLY1 DEFICIENCY KNOWLEDGE GRAPH
Reviewing knowledge as a graph allows exploration of connections between entities otherwise difficult to query in one step and from only one endpoint. The user can explore knowledge and check the supporting evidence. A graph allow to connect diverse type of nodes and relationships, different domains.


=== MORE QUERIES
There are more examples of queries that can be done from traversing the graph looking for paths connecting entities to transcriptome review-expanded summaries. The user can see more examples in notebooks stored in the https://github.com/NuriaQueralt/ngly1-graph/tree/master/neo4j-graphs/ngly1-v3.2/cypher-queries[GitHub repository project - queries]. These notebooks are written in Python programming language and provide an alternative interface to query the graph that allows to tracking queries and results. 

For advanced exploration of the graph please, run the following command by clicking the box to access the advanced guide tutorial. 

[source,cypher]
----
:play http://100.25.145.12:8001/guides/advanced.html
----

NOTE:  The user should have installed https://jupyter.org/[the Jupyter Notebook] software and the IPython kernel for Python3 to run and visualize the queries provided. 


IMPORTANT:  Click the top right **black&white** arrow to go to the next page




